src/rt/sections_elf_shared.d

```diff
@@ -711,12 +711,12 @@ void scanSegments(in ref dl_phdr_info info, DSO* pdso)
         case PT_LOAD:
             if (phdr.p_flags & PF_W) // writeable data segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._gcRanges.insertBack(beg[0 .. phdr.p_memsz]);
             }
             version (Shared) if (phdr.p_flags & PF_X) // code segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._codeSegments.insertBack(beg[0 .. phdr.p_memsz]);
```

Comment:
> Only the address space between [dlpi_addr .. p_vaddr+p_memsz] is accessible. Attempting to read outside this results in a segfault (at least, in our hardened CI environment)

Agreed. However, the fix to align the `_codeSegments[0]` address rounds down. So theoretically this could already cause a out of bounds read. In reality the real start of the section / end of a protected memory segment is probably aligned, so rounding down should be safe. This increases the size of the memory segment though, and increasing the length by the amount of these extra bytes will only cause the  `_codeSegments[$-1]` address to match the real end of the section. So this shouldn't be out of bounds.

> The end address should be on a pointer aligned boundary, and if it isn't, we don't want to read it - else we'll end up segfaulting in the GC.

So I guess the GC simply casts the `_codeSegments` to `size_t[]` as indicated by your example? If this is the case we should indeed not include the trailing bytes. And as we always miss less than `size_t.sizeof` it can't contain a pointer either. So your code should be fine indeed.